File: src/main/java/de/vincentschweiger/phantomclient/modules/UIModule.java
package de.vincentschweiger.phantomclient.modules;

import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import de.vincentschweiger.phantomclient.Mod;
import de.vincentschweiger.phantomclient.modules.settings.impl.DoubleSetting;
import de.vincentschweiger.phantomclient.modules.settings.impl.Setting;
import net.minecraft.client.MinecraftClient;
import net.minecraft.client.util.math.MatrixStack;
import net.minecraft.text.LiteralText;
import net.minecraft.text.Style;
import org.spongepowered.include.com.google.gson.JsonElement;

import java.awt.*;
import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

public abstract class UIModule extends Module {

    private double x, y;
    private int state;
    private int maxState;
    public MatrixStack stack = new MatrixStack();

    public UIModule() {
    }

    public void render(boolean enabled) {
        stack.push();
        if (enabled) {
            MinecraftClient.getInstance().textRenderer.drawWithShadow(stack, getText(), (float) this.getX(), (float) this.getY(), getColor().getRGB());
        } else {
            //If you don't want to render the disabled modules in the drag-screen, just remove/comment following line
            MinecraftClient.getInstance().textRenderer.drawWithShadow(stack, new LiteralText(getText()).setStyle(Style.EMPTY.withStrikethrough(true)), (float) this.getX(), (float) this.getY(), getColor().getRGB());
        }
        stack.pop();
    }

    public String getText() {
        return "";
    }

    public Color getColor() {
        return Color.WHITE;
    }

    public double getX() {
        return x * MinecraftClient.getInstance().getWindow().getScaledWidth();
    }

    public double getY() {
        return y * MinecraftClient.getInstance().getWindow().getScaledHeight();
    }

    public void setPosition(double x, double y) {
        this.x = x / MinecraftClient.getInstance().getWindow().getScaledWidth();
        this.y = y / MinecraftClient.getInstance().getWindow().getScaledHeight();
    }

    public int getWidth() {
        return MinecraftClient.getInstance().textRenderer.getWidth(getText());
    }

    public int getHeight() {
        return 9;
    }

    public void save() {
        settings.put("x", new DoubleSetting(getX()));
        settings.put("y", new DoubleSetting(getY()));
        super.save();
    }


    public void load() {
        super.load();
        double x = ((DoubleSetting) settings.get("x")).get();
        double y = ((DoubleSetting) settings.get("y")).get();
        setPosition(x, y);
    }

    public int getState() {
        return state;
    }

    public void incrState() {
        if (this.state < this.maxState) this.state++;
        else this.state = 0;
    }

    public int getMaxState() {
        return maxState;
    }
}

File: src/main/kotlin/de/vincentschweiger/phantomclient/modules/UIModule.kt
package de.vincentschweiger.phantomclient.modules

import de.vincentschweiger.phantomclient.event.OverlayRenderEvent
import de.vincentschweiger.phantomclient.event.handler
import net.minecraft.client.MinecraftClient
import net.minecraft.client.util.math.MatrixStack
import net.minecraft.text.LiteralText
import net.minecraft.text.Style
import java.awt.Color

abstract class UIModule : Module() {
    private var x = 0.0
    private var y = 0.0
    var state = 0
        private set
    open val maxState = 0
    var stack = MatrixStack()

    val renderHandler = handler<OverlayRenderEvent> {
        render(isEnabled)
    }

    fun render(enabled: Boolean) {
        stack.push()
        if (enabled) {
            MinecraftClient.getInstance().textRenderer.drawWithShadow(stack, text, getX().toFloat(), getY().toFloat(), color.rgb)
        } else {
            //If you don't want to render the disabled modules in the drag-screen, just remove/comment following line
            MinecraftClient.getInstance().textRenderer.drawWithShadow(stack, LiteralText(text).setStyle(Style.EMPTY.withStrikethrough(true)), getX().toFloat(), getY().toFloat(), color.rgb)
        }
        stack.pop()
    }

    open val text: String?
        get() = ""
    open val color: Color
        get() = Color.WHITE

    fun getX(): Double {
        return x * MinecraftClient.getInstance().window.scaledWidth
    }

    fun getY(): Double {
        return y * MinecraftClient.getInstance().window.scaledHeight
    }

    fun setPosition(x: Double, y: Double) {
        this.x = x / MinecraftClient.getInstance().window.scaledWidth
        this.y = y / MinecraftClient.getInstance().window.scaledHeight
    }

    val width: Int
        get() = MinecraftClient.getInstance().textRenderer.getWidth(text)
    val height: Int
        get() = 9

    override fun load() {
        super.load()
        setPosition(0.0, 0.0)
    }

    fun incrState() {
        if (state < maxState) state++ else state = 0
    }
}